src/vs/workbench/services/editor/browser/editorOverrideService.ts

```diff
@@ -40,18 +43,37 @@ export class EditorOverrideService extends Disposable implements IEditorOverride
 	readonly _serviceBrand: undefined;
 
 	private _contributionPoints: Map<string | glob.IRelativePattern, ContributionPoints> = new Map<string | glob.IRelativePattern, ContributionPoints>();
+	private static readonly overrideCacheStorageID = 'editorOverrideService.cache';
+	private cache: Set<string> | undefined;
 
 	constructor(
 		@IEditorGroupsService private readonly editorGroupService: IEditorGroupsService,
 		@IConfigurationService private readonly configurationService: IConfigurationService,
 		@IQuickInputService private readonly quickInputService: IQuickInputService,
 		@INotificationService private readonly notificationService: INotificationService,
 		@ITelemetryService private readonly telemetryService: ITelemetryService,
+		@ILifecycleService private readonly lifeCycleService: ILifecycleService,
+		@IStorageService private readonly storageService: IStorageService,
+		@IExtensionService private readonly extensionService: IExtensionService
 	) {
 		super();
+		// Read in the cache on statup
+		this.cache = new Set<string>(JSON.parse(this.storageService.get(EditorOverrideService.overrideCacheStorageID, StorageScope.GLOBAL, JSON.stringify([]))));
+		this.storageService.remove(EditorOverrideService.overrideCacheStorageID, StorageScope.GLOBAL);
+
+		this._register(this.lifeCycleService.onWillShutdown(() => {
+			// We want to store the glob patterns we would activate on, this allows us to know if we need to await the ext host on startup for opening a resource
+			this.cacheContributionPoints();
+		}));
 	}
 
 	async resolveEditorOverride(editor: IEditorInput, options: IEditorOptions | ITextEditorOptions | undefined, group: IEditorGroup): Promise<IEditorInputWithOptionsAndGroup | undefined> {
+		// If it was an override before we await for the extensions to activate and then proceed with overriding or else they won't be registered
+		if (this.cache && editor.resource && this.resourceMatchesCache(editor.resource)) {
+			await this.extensionService.whenInstalledExtensionsRegistered();
+			this.cache = undefined;
```

Comment:
The lifecycle of `this.cache` is not entirely clear to me. Shouldn't there be some code that always ensures `this.cache` becomes `undefined` when `whenInstalledExtensionsRegistered` and here we only do the lookup?

Reply:
I added a listener to the `onDidRegisterExtensions` to also set the cache to undefined therefore we won't call await for it unless we're the first ones needing registered extensions

Reply:
I would think that the `this.cache = undefined;` can be removed because we now call it above `onDidRegisterExtensions`?